### PR TITLE
Clean up and refactor the output of the OTX CLI

### PR DESCRIPTION
**Pull request by harimkang**
### Summary

Clean up OTX's unorganized output and policies, and various arguments related to output.
- New feature
  - New args in CLI
    - AS-IS (otx v1.1) -> TO-BE (this PR)
      - `--work-dir` -> `--workspace` : otx workspace path.
      - `--save-model-to` & `--save-performance` -> `--output` : output directory path.

  - New output
image

+ added some feedbacks
    - Remove symlink for export & optimize: This can create unnecessary links outside of the workspace.
    - Raise error when otx deploy runs without --load-weights: because the exported model is no longer found in latest
    - Rename output: openvino_models -> openvino & latest -> latest_trained_model
    - Link parent path of otx train result output to latest_trained_model : Link models only -> Link models & logs
image

### How to test

```
# auto-config or train with args
(otx) otx train --train-data-roots data/path

# or build & train
(otx) otx build --task classification
(otx) otx {train, eval, export, optimize}
```

### Checklist

- [x] I submit my changes into the `develop` branch
- [ ] I have added description of my changes into [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md)
- [x] I have updated the [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) accordingly
- [ ] I have added tests to cover my changes
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)

### License

- [ ] I submit _my code changes_ under the same [MIT License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2023 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

### CODE REVIEW

**Comment by harimkang** [comment]
I'm running an E2E test and will attach the results when I get them.

File: otx/cli/manager/config_manager.py [hunk 1]
```diff
@@ -115,6 +117,20 @@ def data_config_file_path(self) -> Path:
...
+        """The path of output directory for workspace.
+
+        Returns:
+            Path: Path of output directory.
+        """
+        if "output" in self.args and self.args.output:
+            output_path = Path(self.args.output)
+        else:
+            output_path = self.workspace_root / "outputs" / self.create_date
+        output_path.mkdir(exist_ok=True, parents=True)
```
Comment (eunwoosh) [inline otx/cli/manager/config_manager.py]:
```suggestion
        if not output_path.exists():
            output_path.mkdir(exist_ok=True, parents=True)
```
To consider a case where the function is called multiple times, how about changing code like this?

File: docs/source/guide/get_started/quick_start_guide/cli_commands.rst [hunk 2]
```diff
@@ -150,7 +150,7 @@ Training
 - ``weights.pth`` - a model snapshot
 - ``label_schema.json`` - a label schema used in training, created from a dataset
 
-The results will be saved in ``./model`` folder by default. The output folder can be modified by ``--save-model-to`` option. These files are used by other commands: ``export``, ``eval``, ``demo``, etc.
+The results will be saved in ``./model`` folder by default. The output folder can be modified by ``--output`` option. These files are used by other commands: ``export``, ``eval``, ``demo``, etc.
```
Comment (sungmanc) [inline docs/source/guide/get_started/quick_start_guide/cli_commands.rst]:
```suggestion
The results will be saved in ``./outputs`` folder by default. The output folder can be modified by ``--output`` option. These files are used by other commands: ``export``, ``eval``, ``demo``, etc.
```

Comment (sungmanc) [reply docs/source/guide/get_started/quick_start_guide/cli_commands.rst]: In my understanding, the final output is made in `outputs`. and there are `2023000_00000/logs`, `2023000_00000/models` in the `outputs` folder. So, above statement needs to be changed

**Review by sungmanc — COMMENTED** [review]
Overall LGTM, this is the pretty good improvement :) 

BTW, for the deploy, eval, export, optimize, and train.py, will we support `-o` argument?
In the documentation, there is no description for `-o` argument. So, how about removing this argument?

+ We have 3 scenarios as below, how about adding simple unit-tests for these scenarios?

1. --output  (no --workspace)
2. --workspace (no --output)
3. --workspace + --output

File: docs/source/guide/get_started/quick_start_guide/cli_commands.rst (hunk 2 above)
Comment (harimkang) [reply docs/source/guide/get_started/quick_start_guide/cli_commands.rst]: Yes, I agree. thanks! I'll fix it, and also try to add unit-test (maybe should add it to integration/cli test). I think `-o` would be nice to have for convenience, what do you think? I can fix the documentation, or I'll delete it as you suggest.

Comment (sungmanc) [reply docs/source/guide/get_started/quick_start_guide/cli_commands.rst]: Editing documentation is also an excellent way.

**Comment by eunwoosh** [comment]
> Overall LGTM, this is the pretty good improvement :)
> 
> BTW, for the deploy, eval, export, optimize, and train.py, will we support `-o` argument? In the documentation, there is no description for `-o` argument. So, how about removing this argument?
> 
> * We have 3 scenarios as below, how about adding simple unit-tests for these scenarios?
> 
> 1. --output  (no --workspace)
> 2. --workspace (no --output)
> 3. --workspace + --output

IMHO, `-o` is also good argument to use. How about to update documentation instead of removing it?

File: docs/source/guide/get_started/quick_start_guide/cli_commands.rst (hunk 2 above)
Comment (harimkang) [reply docs/source/guide/get_started/quick_start_guide/cli_commands.rst]: Yeah, I'm only going to modify the --help output part of the cli_commands part, I think the example can be left as is.
Thanks for comment :)

File: otx/cli/tools/build.py [hunk 3]
```diff
@@ -91,13 +91,13 @@ def main():
     config_manager = ConfigManager(args, mode="build")
     if args.task:
         config_manager.task_type = args.task.upper()
-    if args.work_dir:
-        config_manager.workspace_root = Path(args.work_dir)
+    if args.workspace:
```
Comment (eunwoosh) [inline otx/cli/tools/build.py]: Is this code necessary? As far as I understand, init function of `ConfigManager` class set it by itself.

File: otx/cli/tools/deploy.py [hunk 4]
```diff
@@ -82,11 +82,10 @@ def main():
 
     deployed_model = ModelEntity(None, environment.get_model_configuration())
 
-    if "save_model_to" not in args or not args.save_model_to:
-        args.save_model_to = str(config_manager.workspace_root / "model-deployed")
-    os.makedirs(args.save_model_to, exist_ok=True)
+    output_path = Path(args.output) if args.output else config_manager.output_path
```
Comment (eunwoosh) [inline otx/cli/tools/deploy.py:82]: how about using `config_manager.output_path`? Same algorithm exists there.

File: otx/cli/tools/eval.py [hunk 5]
```diff
@@ -136,9 +137,8 @@ def main():
     assert resultset.performance is not None
     print(resultset.performance)
 
-    if not args.save_performance:
-        args.save_performance = str(Path(args.load_weights).parent / "performance.json")
-    with open(args.save_performance, "w", encoding="UTF-8") as write_file:
+    output_path = Path(args.output) if args.output else config_manager.output_path
```
Comment (eunwoosh) [inline otx/cli/tools/eval.py:143]: how about just using `config_manager.output_path`? Same algorithm exists there.

File: otx/cli/tools/build.py (hunk 3 above)
Comment (harimkang) [reply otx/cli/tools/build.py]: In this code, workspace_root is not passed in init.

File: otx/cli/manager/config_manager.py (hunk 1 above)
Comment (harimkang) [reply otx/cli/manager/config_manager.py]: I fixed this. https://github.com/openvinotoolkit/training_extensions/pull/1946/commits/2cb8133579f0f6a36f78f4a32ef2a5287d1008e8

File: otx/cli/tools/build.py (hunk 3 above)
Comment (harimkang) [reply otx/cli/tools/build.py]: I fixed this. https://github.com/openvinotoolkit/training_extensions/pull/1946/commits/2cb8133579f0f6a36f78f4a32ef2a5287d1008e8

File: otx/cli/tools/deploy.py (hunk 4 above)
Comment (harimkang) [reply otx/cli/tools/deploy.py:82]: I fixed this. https://github.com/openvinotoolkit/training_extensions/pull/1946/commits/2cb8133579f0f6a36f78f4a32ef2a5287d1008e8

File: otx/cli/tools/eval.py (hunk 5 above)
Comment (harimkang) [reply otx/cli/tools/eval.py:143]: I fixed this. https://github.com/openvinotoolkit/training_extensions/pull/1946/commits/2cb8133579f0f6a36f78f4a32ef2a5287d1008e8

**Comment by harimkang** [comment]
We decided to move the `latest` folder to `outputs`.
I'm also going to add an integration test to check the CLI behavior with or without `--output` or `--workspace` and test for file creation checking.

This is currently working as expected, so I'm going to leave the PR open for reviews, and I'd love to hear your thoughts on this PR. :)

File: otx/algorithms/common/tasks/training_base.py [hunk 6]
```diff
@@ -190,6 +190,11 @@ def project_path(self):
         """Return output path with logs."""
         return self._output_path
 
+    @property
+    def config(self):
```
Comment (jaegukhyun) [inline otx/algorithms/common/tasks/training_base.py:194]: I think the name "config" is too broad given that this is only using for mm training backend. 
How about final_recipe or final_config?

Comment (harimkang) [reply otx/algorithms/common/tasks/training_base.py:194]: First of all, anomaly is also using an attribute called **config**. I've named it as a common attribute name, which seems to be up for discussion.

**Review by jaegukhyun — DISMISSED** [review]
I looked only code parts(it's really hard to look docs and test files...), and the code works look really good to me. Thank you for your effort

File: docs/source/guide/get_started/quick_start_guide/cli_commands.rst (hunk 2 above)
Comment (harimkang) [reply docs/source/guide/get_started/quick_start_guide/cli_commands.rst]: i added integration test cases related this PR https://github.com/openvinotoolkit/training_extensions/pull/1946/commits/673f268f937df627cb15aa06f4153461c965d882

**Review by jaegukhyun — DISMISSED** [review]

File: tests/integration/cli/test_cli.py [hunk 7]
```diff
@@ -120,6 +121,122 @@ def test_otx_train(self, case, tmp_dir_path):
...
+        # check output (without --output -> Default outputs)
+        output_path = os.path.join(tmp_dir_path, "otx-workspace", "outputs")
+        assert os.path.exists(output_path)
+        file_list = sorted(os.listdir(output_path))
+        assert len(file_list) == 2
+        assert os.path.exists(os.path.join(output_path, "latest_trained_model"))
+        assert os.path.exists(os.path.join(output_path, "latest_trained_model", "models"))
+        assert os.path.exists(os.path.join(output_path, "latest_trained_model", "models", "weights.pth"))
+        assert os.path.exists(os.path.join(output_path, "latest_trained_model", "models", "label_schema.json"))
+        file_list = file_list[:-1]  # Remove latest from list, then file_list[-1] is latest
```
Comment (eunwoosh) [inline tests/integration/cli/test_cli.py]:
```suggestion
        file_list.remove("latest_trained_model")
```

File: tests/integration/cli/test_cli.py [hunk 8]
```diff
@@ -120,6 +121,122 @@ def test_otx_train(self, case, tmp_dir_path):
...
+            os.path.join(tmp_dir_path, "otx-workspace"),
+        ]
+        check_run(command_line)
+
+        # check output (without --output -> Default outputs)
+        output_path = os.path.join(workspace_path, "outputs")
+        assert os.path.exists(output_path)
+        file_list = sorted(os.listdir(output_path))
+        assert len(file_list) == 3
+        file_list = file_list[:-1]  # Remove latest from list, then file_list[-1] is latest
```
Comment (eunwoosh) [inline tests/integration/cli/test_cli.py]:
```suggestion
        file_list.remove("latest_trained_model")
```

File: tests/integration/cli/test_cli.py [hunk 9]
```diff
@@ -120,6 +121,122 @@ def test_otx_train(self, case, tmp_dir_path):
...
+            os.path.join(tmp_dir_path, "otx-workspace"),
+        ]
+        check_run(command_line)
+
+        # check output (without --output -> Default outputs)
+        output_path = os.path.join(workspace_path, "outputs")
+        assert os.path.exists(output_path)
+        file_list = sorted(os.listdir(output_path))
+        assert len(file_list) == 4
+        file_list = file_list[:-1]  # Remove latest from list, then file_list[-1] is latest
```
Comment (eunwoosh) [inline tests/integration/cli/test_cli.py]:
```suggestion
        file_list.remove("latest_trained_model")
```

File: tests/integration/cli/test_cli.py [hunk 10]
```diff
@@ -120,6 +121,122 @@ def test_otx_train(self, case, tmp_dir_path):
...
+        assert os.path.exists(expected_workspace_path)
+        expected_output_path = os.path.join(expected_workspace_path, "outputs")
+        assert os.path.exists(expected_output_path)
+        file_list = sorted(os.listdir(expected_output_path))
+        assert len(file_list) == 2
+        assert os.path.exists(os.path.join(expected_output_path, "latest_trained_model"))
+        assert os.path.exists(os.path.join(expected_output_path, "latest_trained_model", "models"))
+        assert os.path.exists(os.path.join(expected_output_path, "latest_trained_model", "models", "weights.pth"))
+        assert os.path.exists(os.path.join(expected_output_path, "latest_trained_model", "models", "label_schema.json"))
+        file_list = file_list[:-1]  # Remove latest from list, then file_list[-1] is latest
```
Comment (eunwoosh) [inline tests/integration/cli/test_cli.py]:
```suggestion
        file_list.remove("latest_trained_model")
```

**Review by eunwoosh — APPROVED** [review]

**Review by sungmanc — APPROVED** [review]

**Pull request merged**